### PR TITLE
fix: offline recording crash

### PR DESCRIPTION
**Pull request by Rolph1**
Due to issues with merge conflicts some of the logic to save the files was modified and causes crashes.
The exact issue is: When recording sometimes the file is sometimes wrongly saved under the name "audio_record.wav" and so when trying to play it in the review offline recording screen or offline recording feedback screen the app would crash.

The issue is due to: SpeakingViewModel using the function startRecording() which would by default save the values to "audio_record.wav"
How it was fixed: A new function startRecording(audioFile: File) has been added to AudioRecorder and AudioRecorderRepository is now called to save the file with the correct file name.

### **Fix for File Naming Issue in Offline Recording**

**Problem:**
The app crashes during playback in the review or feedback screen because the recorded file is sometimes saved with the default name `audio_record.wav`, causing a mismatch when the correct file name is expected (which is a random string generated per interview to allow the device to correctly store a lot of interviews).

**Root Cause:**
The `SpeakingViewModel` uses the `startRecording()` function, which defaults to saving files as `audio_record.wav`. This leads to conflicts and crashes during playback since the expected file name differs.

**Solution:**
1. Introduced a new function `startRecording(audioFile: File)` in `AudioRecorder` to manually select the file name.
2. Updated `AudioRecorderRepository` and `SpeakingViewModel` to use this function, ensuring the correct file name is used consistently.

**Key Changes:**
- Refactored `AudioRecorder` and `SpeakingRepositoryRecord` to accept a `File` parameter for saving.
- Updated UI components (`OfflineRecordingScreen`, `RecordingReviewScreen`) to pass the correct file name.
- Now creating one context val in MainActivity and passing it to every screen instead of calling LocalContext.current to make code more readable and since I thought it might have caused issues with the recorder (it didn't in the end)
---

### **Files Changed**

1. **`AudioRecorder.kt`:**
   - Added a new method `startRecording(audioFile: File)` to allow saving files with a specific name.
   - Updated logging to verify successful file saving with the correct name.

2. **`SpeakingRepositoryRecord.kt`:**
   - Replaced the call to `startRecording()` with the new `startRecording(audioFile: File)` method.
   - Ensured that the passed file name is consistent across components.

3. **`SpeakingViewModel.kt`:**
   - Added support for passing a `File` object when starting or saving a recording.
   - Modified `onMicButtonClicked` to pass the correct file name to `SpeakingRepository`.

4. **`OfflineRecordingScreen.kt`:**
   - Updated the logic for initializing the file name (`{interviewPromptNb}.mp3`).
   - Passed the generated file name to the recording logic to ensure consistency.

5. **`RecordingReviewScreen.kt`:**
   - Ensured the correct file name is used when attempting to play back a recording.
   - Added logging to verify the playback file matches the saved file.

---

### **Outcome**
- Fixed crashes caused by file name mismatches during playback.
- Improved consistency in file naming across the app.

### CODE REVIEW

**Review by baptisteghostine — CHANGES_REQUESTED** [review]
#### **Summary**
This PR addresses a critical issue where recordings in the offline recording flow were sometimes saved under a default filename (`audio_record.wav`) instead of a unique name. This mismatch caused crashes during playback, especially in the review screens where the correct file name was expected. The PR introduces a parameterized approach to saving recordings, ensuring consistent and correct file naming throughout the application.

---

#### **What’s Good**
- **Clear Problem Identification**: The PR clearly states the original problem (conflicts due to a default filename) and the resulting crashes. The motivation for these changes is easy to understand.
- **Well-Explained Solution**: Introducing a method in `AudioRecorder` to accept a `File` object ensures file naming is explicit and consistent. This eliminates reliance on a default filename in `SpeakingViewModel` and `SpeakingRepositoryRecord`.
- **Refactoring for Clarity**: Passing the `File` object through `SpeakingViewModel` and `OfflineRecordingScreen` ensures there’s a single source of truth for the filename, preventing similar issues in the future.
- **Improved Logging**: The addition of logging for successful file creation provides useful debugging information and helps verify correct file handling.

---

#### **Suggestions for Improvement**
- **Testing Additional Scenarios**:
  - While the updated tests verify the WAV header and recording stopping functionality, consider adding a test to ensure that providing a custom `File` name results in the correct filename being used. This scenario-specific test would further validate the changes.
- **Documentation Comments**:
  - Add a brief comment in the code to explain the introduction of `saveRecording(audioFile: File)` and its purpose. This will help maintainers quickly understand the reason behind this approach.
- **Error Handling**:
  - Ensure there is sufficient error handling for cases where file writing might fail or if the file already exists. While the problem appears resolved, adding a safeguard (such as additional logging) would help prevent or diagnose similar issues in the future.

---

#### **Overall Assessment**
This PR is a well-thought-out solution that correctly identifies the root cause and implements a direct fix. By introducing a method to save recordings with a specified filename and ensuring that filename is consistently passed along, the reliability of the offline recording process is greatly improved. The explanation is thorough, and the code adjustments make logical sense and align with the stated goals.

---

#### **Conclusion**
**Waiting for CI to pass**. The PR resolves the crash issue and improves code clarity around file handling. The changes appear stable and beneficial for the long-term maintainability of the offline recording feature. With the suggested enhancements, this PR will be even stronger. I'm waiting for the PR to pass before approving changes.

File: app/src/main/java/com/github/se/orator/model/symblAi/AudioRecorder.kt
Comment (baptisteghostine) [inline app/src/main/java/com/github/se/orator/model/symblAi/AudioRecorder.kt:103]: Add a brief comment above the saveAsWavFile((audioData: ByteArray, audioFile: File)) method to explain why this overload was introduced. I started writing a comment but use the /** ... **/ for writing comments.

```
// This overload of startRecording() allows specifying a custom audio file name.
// It was introduced to ensure that the recorded audio can be saved under a unique, 
// caller-defined filename, preventing conflicts and issues in playback due to default names.
fun startRecording(audioFile: File = File(context.cacheDir, "audio_record.wav")) {
    // Existing code...
}
```

File: app/src/test/java/com/github/se/orator/model/symblAi/AudioRecorderTest.kt
Comment (baptisteghostine) [inline app/src/test/java/com/github/se/orator/model/symblAi/AudioRecorderTest.kt:106]: You could also add a test to verify that providing a custom File object to startRecording(audioFile: File) indeed results in the recording being saved under that specific file name. However, this is not mandatory for now and could be done in next sprint.

**Review by baptisteghostine — CHANGES_REQUESTED** [review]
I added comments to the code with the requested changes

File: app/src/main/java/com/github/se/orator/model/symblAi/SpeakingRepositoryRecord.kt
Comment (Rolph1) [reply app/src/main/java/com/github/se/orator/model/symblAi/SpeakingRepositoryRecord.kt:26]: Great idea! I haven't done this since I overrode an already existing method that didn't have this feature. Another PR to make the repository more secure should be done in the future.

File: app/src/main/java/com/github/se/orator/model/symblAi/AudioRecorder.kt
Comment (Rolph1) [reply app/src/main/java/com/github/se/orator/model/symblAi/AudioRecorder.kt:103]: This method was not overloaded, are you sure you're talking about the right one?

File: app/src/test/java/com/github/se/orator/model/symblAi/AudioRecorderTest.kt
Comment (Rolph1) [reply app/src/test/java/com/github/se/orator/model/symblAi/AudioRecorderTest.kt:106]: Good idea, I'll implement this in the next commit.

**Comment by baptisteghostine** [comment]
Hi Rudolf, thank you for resolving the requested changes. It looks great! 

However, I can see that the line coverage is quite low. 

image

Could you explain the reason for this?

After that I will accept the PR. Again, I appreciate your great work, keep going!

**Comment by Rolph1** [comment]
> Hi Rudolf, thank you for resolving the requested changes. It looks great!
> 
> However, I can see that the line coverage is quite low.
> 
> image
> 
> Could you explain the reason for this?
> 
> After that I will accept the PR. Again, I appreciate your great work, keep going!



> Hi Rudolf, thank you for resolving the requested changes. It looks great!
> 
> However, I can see that the line coverage is quite low.
> 
> image
> 
> Could you explain the reason for this?
> 
> After that I will accept the PR. Again, I appreciate your great work, keep going!

Hi Baptiste, a lot of the uncovered code are things that can't directly be tested with Mockito such as logs or variables that have been declared:
image
I looked a round a bit more and found ~5 lines out of 18 that could directly be tested but that's about it. I just added a lot of logs to debug which can't be tested! 
Also since it's now about 13 lines of untested code I think it's okay for now.

Cheers

**Comment by baptisteghostine** [comment]
Ok! I agree, you've already done your best to maximize coverage and we are already above the minimal coverage overall. 

I will then accept the PR. Thanks for your work!

**Review by baptisteghostine — APPROVED** [review]

**Pull request merged**
